File: src/cpp_pubsub_srv/src/talker_adv.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "custom_interfaces/msg/location.hpp"
#include "custom_interfaces/srv/authenticate_listener.hpp"

using namespace std::chrono_literals;

/* This example creates a subclass of Node and uses std::bind() to register a
* member function as a callback from the timer. */

class AdvancedTalker : public rclcpp::Node
{
  public:
    AdvancedTalker()
    : Node("adv_talker"), password(12345)
    {
      publisher_ = this->create_publisher<custom_interfaces::msg::Location>("custom_topic_adv", 10);
      timer_ = this->create_wall_timer(
      500ms, std::bind(&AdvancedTalker::timer_callback, this));

      location_service_ = this->create_service<custom_interfaces::srv::AuthenticateListener>(
      "/adv_talker/share_secret_location",
      std::bind(&AdvancedTalker::share_location_request, this, std::placeholders::_1, std::placeholders::_2)
    );

    }

  private:
    void timer_callback()
    {
      auto msg = custom_interfaces::msg::Location();
      msg.location.x = 1.0;
      msg.location.y = 1.0;
      msg.location.z = 0.0;
      msg.name = "honolulu";

      publisher_->publish(msg);
    }
    
     void share_location_request(
      const std::shared_ptr<custom_interfaces::srv::AuthenticateListener::Request> request,
      std::shared_ptr<custom_interfaces::srv::AuthenticateListener::Response> response)
    {
      
      if (request->password == this->password) {
        response->secret_location.location.x = 20.0;
        response->secret_location.location.y = 15.0;
        response->secret_location.location.z = 5.0;
        response->secret_location.name = "Haiti";

        RCLCPP_INFO(this->get_logger(), "Authentication successful. Returning location.");
      } else {
        response->secret_location.location.x = 0.0;
        response->secret_location.location.y = 0.0;
        response->secret_location.location.z = 0.0;
        response->secret_location.name = "unauthorized";

        RCLCPP_WARN(this->get_logger(), "Authentication failed. Wrong password: %ld", request->password);
      }
    }


    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<custom_interfaces::msg::Location>::SharedPtr publisher_;
    rclcpp::Service<custom_interfaces::srv::AuthenticateListener>::SharedPtr location_service_;
    int password;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<AdvancedTalker>());
  rclcpp::shutdown();
  return 0;
}

File: src/py_pubsub_srv/py_pubsub_srv/talker_adv.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from custom_interfaces.msg import Location
from custom_interfaces.srv import AuthenticateListener

class AdvancedTalker(Node):

    def __init__(self):
        super().__init__('adv_talker')
        self.password = 12345

        # Publisher
        self.publisher_ = self.create_publisher(Location, 'custom_topic_adv', 10)
        self.timer_ = self.create_timer(0.5, self.timer_callback)

        # Service
        self.location_service_ = self.create_service(
            AuthenticateListener,
            '/adv_talker/share_secret_location',
            self.share_location_request
        )

        self.get_logger().info("AdvancedTalker node started.")

    def timer_callback(self):
        msg = Location()
        msg.location.x = 1.0
        msg.location.y = 1.0
        msg.location.z = 0.0
        msg.name = "honolulu"
        self.publisher_.publish(msg)
        self.get_logger().info("Published location: (1.0, 1.0, 0.0) / honolulu")

    def share_location_request(self, request, response):
        if request.password == self.password:
            response.secret_location.location.x = 20.0
            response.secret_location.location.y = 15.0
            response.secret_location.location.z = 5.0
            response.secret_location.name = "Haiti"
            self.get_logger().info("Authentication successful. Returning location.")
        else:
            response.secret_location.location.x = 0.0
            response.secret_location.location.y = 0.0
            response.secret_location.location.z = 0.0
            response.secret_location.name = "unauthorized"
            self.get_logger().warn(f"Authentication failed. Wrong password: {request.password}")
        return response


def main(args=None):
    rclpy.init(args=args)
    node = AdvancedTalker()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info("Shutting down AdvancedTalker node.")
    finally:
        node.destroy_node()
        rclpy.shutdown()